### PR TITLE
DrivAerML: 4L/320d + cosine T_max=20 (more restarts)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Reducing cosine annealing T_max from 30 to 20 gives 50% more learning rate restarts within the 180-min wall-clock budget. More restarts help the optimizer escape local minima and explore the loss landscape more thoroughly. The 4L/320d configuration already achieves 5.027%, and tighter restarts may push it further by allowing finer convergence near the end of each cycle.

## Instructions
Run from `target/icml2026` with the following command (copy exactly):

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 20 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 320 \
  --model-heads 5 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-4L320d-tmax
```

The only change from baseline is `--cosine-t-max 20` (was 30). Do not change any other hyperparameters.

## Baseline
Current DrivAerML best (PR #2648, W&B run `qx7z7if3`):
- `val_primary/surface_rel_l2_pct`: **5.027%** (val) / 6.244% (test)
- Epoch: 257 of 180-min budget
- External competition target: **< 3.71%**

Baseline reproduce command:
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

Beat `val_primary/surface_rel_l2_pct < 5.027%` to be a winner.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT ymir:

## Results

**Experiment:** DrivAerML 4L/320d + T_max=20 (Fourier, no-EMA)

| Metric | Baseline (PR #2648, 4L/320d T_max=30, 257ep) | This Run |
|--------|-----------------------------------------------|----------|
| val_primary/surface_rel_l2_pct | 5.027% | **10.27%** (best, epoch 76) |
| test_primary/surface_rel_l2_pct | 6.244% | 46.74% (diverged model) |
| Epochs | 257 | 269 |

**NEGATIVE RESULT — experiment failed due to catastrophic training instability.**

### Command
```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 20 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 320 \
  --model-heads 5 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent ymir --wandb-name "ymir/drivaerml-4L320d-tmax20" \
  --wandb-group "drivaerml-4L320d-tmax"
```

### W&B
- Run ID: `ofecgd5w`
- URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ofecgd5w

### Peak Memory
- 8 GPUs used (DDP), batch_size=1, 50k surface points

### What happened

The model trained well initially, reaching **10.27% val** at epoch 76 (approaching but not beating the 5.027% baseline). However, the training suffered **catastrophic gradient explosions** starting at epoch 78:

**Timeline of instability:**
- Epochs 1-76: Normal training. Val improved steadily: 60.2% → 30.6% → 15.5% → 12.8% → 10.6% → **10.27%** (best)
- Epoch 78: First major spike — val jumped to 37.71%, grad_norm to 1.97
- Epoch 79: Val 40.86%, grad_norm 9.0
- Epochs 80-96: Partial recovery to ~16-18%, but grads remained elevated (3-5x normal)
- Epoch 97: **Second catastrophic divergence** — val 64.68%, grad_norm 44.6
- Epochs 98-130: Oscillating 16-27%, never recovering to pre-crash levels
- Epoch 163: **Terminal explosion** — grad_norm hit 3.65 million, val 56%
- Epochs 163-269: Grad norms escalated exponentially: millions → billions → trillions (peak 128 trillion at epoch 268). Model weights destroyed.

**Key observations:**
1. The instability is **not LR-driven** — crashes occurred at LR=0 (cycle bottoms) as well as at peak LR
2. The grad_norm grew **exponentially** across cycles, indicating self-reinforcing weight damage
3. No gradient clipping was used — this is almost certainly the root cause
4. The 4L/384d model (PR #2687) with T_max=40 did NOT show this instability, suggesting the smaller 320d model or shorter T_max=20 cycles amplify the problem
5. The final test metric (46.74%) is from the destroyed model and is meaningless

### Suggested follow-ups
- **Gradient clipping is essential** for the 4L/320d architecture — try `--grad-clip 1.0` or `--grad-clip 5.0`
- Compare stability of T_max=20 vs T_max=30 with gradient clipping enabled
- The 4L/384d model showed better stability — the larger hidden dim may provide implicit regularization
- Early stopping or checkpoint-based recovery could salvage runs that hit transient instabilities
- Weight decay tuning may also help prevent the self-reinforcing gradient explosion

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2756 because ADVISOR: Closing — T_max=20 at 4L/320d diverged catastrophically (10.274%, 2.2x worse than baseline 4.619%). Consistent with rei's T_max=20 divergence at 4L/512d. DrivAerML requires T_max≥30 for stability. Shorter cosine cycles create too-aggressive LR peaks for 3D automotive CFD. The winner (4L/512d+T_max=30) uses T_max=30 for good reason. Assigning you 4L/512d with the standard config..
